### ISSUE TITLE
Cannot restore large agent (18GB)

### ISSUE THREAD

**Issue by kwokkawinghk** [tensorforce/tensorforce#576]
Hi,

I have trained a large PPO agaent model (18GB) and try to restore it for further training. However, the below error code occur and system start a new agent to train:

2019-03-16 08:12:32.430456: W tensorflow/core/framework/op_kernel.cc:1401] OP_REQUIRES failed at save_restore_v2_ops.cc:184 : Out of range: Read fewer bytes than requested

The below is the code that I used to restore the agent, it works fine when the model is below 2GB:

    try :
        agent.restore_model("./model_new/")
    except:
        agent.save_model("./model_new/")
    return agent

I am using the below system spec:

Google VM 8 CPU 2GHz
512 GB RAM and hard disk
Windows Server 2016
Tensorflow 1.13.1
Tensorforce 0.4.3
Anaconda Python 3.6.8

How should I do to restore a large model? Thank you

**Comment by kwokkawinghk** [comment]
below is my model as attached:

def create_network_spec():
    network_spec = [
        {"type": "flatten"},
        dict(size= 152, type = 'dense', activation = 'selu', l2_regularization = 0.007920429669304596, 
             l1_regularization = 0.07371959121453227),
        dict(type = 'dropout', rate = 0.12911336420147002),
        dict(size= 228, type = 'dense', activation = 'selu', l2_regularization = 0.007920429669304596, 
             l1_regularization = 0.07371959121453227),
        dict(type = 'dropout', rate = 0.12911336420147002),
        dict(type='internal_lstm', size=456, dropout = 0.12911336420147002),
        dict(type='internal_lstm', size=456, dropout = 0.12911336420147002),
        dict(type='internal_lstm', size=456, dropout = 0.12911336420147002),
    ]
    return network_spec

def create_baseline_spec():
    baseline_spec = [
        {
            "type": "lstm",
            "size": 456,
        },
        dict(type='dense', size=228, activation='selu'),
        dict(type = 'dropout', rate = 0.12911336420147002),
        dict(type='dense', size=152, activation='selu'),
        dict(type = 'dropout', rate = 0.12911336420147002),
        dict(type='dense', size=114, activation='selu'),
        dict(type = 'dropout', rate = 0.12911336420147002),
    ]
    return baseline_spec

def testing_agent(environment):
    
    network_spec = create_network_spec()
    baseline_spec = create_baseline_spec()
    agent = PPOAgent(
        discount=0.9677484438906688,
        states=environment.states,
        actions=environment.actions,
        network=network_spec,
        states_preprocessing=None,
        actions_exploration=None,
        reward_preprocessing=None,
        update_mode=dict(
            unit= 'episodes',
            batch_size= 800,
            frequency=12
        ),
        memory=None, 
        distributions=None,
        entropy_regularization=0.8766700482699176,

        baseline_mode='states',
        baseline=dict(type='custom', network=baseline_spec),
        baseline_optimizer=dict(
            type='multi_step',
            optimizer=dict(
                type='adam',
                learning_rate=0.7342086295731822
            ),
            num_steps=17
        ),
        gae_lambda=0.9285969986276199,
        likelihood_ratio_clipping=0.9837192515975488,
        step_optimizer=dict(
            type='adam',
            learning_rate=0.7342086295731822
        ),
        subsampling_fraction=0.1,
        optimization_steps=17,
        execution=None
    )
    try :
        agent.restore_model("./model_new/")
    except:
        agent.save_model("./model_new/")
    return agent

**Comment by AlexKuhnle** [comment]
Are you on the latest GitHub version of the framework, or the PyPI version? If the latter, maybe worth updating to the GitHub version, as pip wasn't updated for a while. Apart from that I'm not aware of a limit for model size (and would be surprised if there was one), but the behavior you describe doesn't sound like a Tensorforce-related issue (if it works well below a certain size). Have you googled about other people experiencing problems with loading models of a larger size?

However, besides that, have you looked into the `update_mode` and `memory` values you chose? As these are responsible for the majority of memory consumption, some explanation around that: `batch_size=800` means you take a batch of 800 episodes (not timesteps!), so depending on how long episodes are, that may be a lot. The default memory size is (arbitrarily) chosen to be `1000 * batch_size`, so if your episodes are shorter, you can save some space there, and you can definitely save space by choosing a smaller `batch_size`. For instance, 80 and even less should be perfectly reasonable and will likely reduce your model size a lot.

**Comment by kwokkawinghk** [comment]
Hi Alex,

Many thanks for your reply. I successfully restore my model after reducing the batch size of my project and it becomes less than 1GB after I changed that to 36 instead of 800. It does not mean that the problem of restoring a large model has been fixed, although I have fixed this problem temporarily.

I have searched on google before raising this issue and this seems that a tensorflow issue and I also reopen the issue #18769 in tensorflow github page. People have suggested that using Variable Partitioning with tf.variable_axis_size_partitioner and max_shard_bytes=2 ** 30 large embeddings loading works well.

I have tried that on the "models/model.py" but failed, so I raised this issue. Hope you can have a look as well. Highly appreciated about your suggestion.

Ka Wing

**Comment by AlexKuhnle** [comment]
No problem, and thanks for the additional information, closing this then.
